frontends/p4/decl_copyprop.h

```diff
@@ -0,0 +1,103 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _P4_DECL_COPYPROP_H_
+#define _P4_DECL_COPYPROP_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/common/resolveReferences/resolveReferences.h"
+
+namespace P4 {
+
+/// DeclarationCopyPropagation pass operates on control locals block and
+/// propagates variables only if they are initialized with constant values.
+/// For example:
+///     ...
+///  control c() {
+///     bit<32> a = 10;
+///     S s = { x = 3, y = 4, z = 5 };
+///     bit<32> b = a;
+///     bit<32> c = s.y;
+///     apply {
+///         /* not processed */
+///     }
+///  }
+/// <----- DeclarationCopyPropagation ----->
+///     ...
+///  control c() {
+///     bit<32> a = 10;
+///     S s = { x = 3, y = 4, z = 5 };
+///     bit<32> b = 10;
+///     bit<32> c = 4;
+///     apply {
+///         /* not processed */
+///     }
+///  }
+
+class DoDeclarationCopyPropagation : public Transform {
+    ReferenceMap* refMap;
+    TypeMap* typeMap;
+
+    /// Map for storing constant values for variables. The mapped value
+    /// is a pointer to an Expression that represents the constant value
+    /// for the variable with the name given as a map key.
+    std::map<cstring, const IR::Expression*> varMap;
+
+    /// Controlls where should the propagation work.
+    bool working = false;
+
+    /// If set to true limits the pass to only check if the used
+    /// variables are initialized. It assumes that all of them are
```

Comment:
When checkOnly is true the pass checks if used variables have declarations, and if they have an initializer sets them as compile time constant, even if they aren't. This is done to postpone compile time constant check and wait for TypeInference to insert casts and convert structures from ListExpression to StructExpression. When this flag is true no propagation is done.
Propagating variables initialized by calling functions is currently disabled because I think that it is too early in the frontend to find the return values and propagate them without complicating things further.